Service/ReplicasClient.php

```diff
@@ -0,0 +1,131 @@
+
+    /** @var ContainerInterface */
+    protected $container;
+
+    /** @var Connection[] Keys are the slice (i.e. 's1') and values are Connection objects. */
+    protected $connections;
+
+    /**
+     * ReplicasClient constructor.
+     * @param HttpClientInterface $httpClient
+     * @param CacheItemPoolInterface $cache
+     * @param ContainerInterface $container
+     */
+    public function __construct(
+        HttpClientInterface $httpClient,
+        CacheItemPoolInterface $cache,
+        ContainerInterface $container
+    ) {
+        $this->httpClient = $httpClient;
+        $this->cache = $cache;
+        $this->container = $container;
+    }
+
+    /**
+     * Fetch and concatenate all the dblists into one array.
+     * @return string[] Keys are the db name (i.e. 'enwiki'), values are the slice (i.e. 's1')
+     */
+    public function getDbList(): array
+    {
+        $cacheItem = $this->cache->getItem('toolforgebundle.dblists');
+        if ($cacheItem->isHit()) {
+            return $cacheItem->get();
+        }
+
+        $dbList = [];
+        $exists = true;
+        $i = 0;
+
+        while ($exists) {
+            $i += 1;
+            $response = $this->httpClient->request('GET', self::DBLISTS_URL."s$i.dblist");
+            $exists = in_array(
+                $response->getStatusCode(),
+                [Response::HTTP_OK, Response::HTTP_NOT_MODIFIED]
+            ) && $i < 50; // Safeguard
+
+            if (!$exists) {
+                break;
+            }
+
+            $lines = explode("\n", $response->getContent());
+            foreach ($lines as $line) {
+                $line = trim($line);
+                if (1 !== preg_match('/^#/', $line) && '' !== $line) {
+                    // Skip comments and blank lines.
+                    $dbList[$line] = "s$i";
+                }
+            }
+        }
+
+        $cacheItem->set($dbList)
+            ->expiresAfter(new DateInterval(self::DBLIST_CACHE_DURATION));
+        $this->cache->save($cacheItem);
+
+        return $dbList;
+    }
+
+    /**
+     * Get a Doctrine Connection instance for the given database.
+     * Any databases that live on the same shard will share the same Connection instance.
+     * @param string $db
+     * @return Connection
+     */
+    public function getConnection(string $db): Connection
+    {
+        $slice = $this->getDbList()[$db];
+        if (isset($this->connections[$slice])) {
+            return $this->connections[$slice];
+        }
+
+        $host = str_replace('*', $slice, $this->container->getParameter('toolforge.replicas.host'));
+
+        $config = [
+            'user' => $this->container->getParameter('toolforge.replicas.username'),
+            'password' => $this->container->getParameter('toolforge.replicas.password'),
+            'host' => $host,
+            'port' => $this->getPortForSlice($slice),
+        ];
+
+        $this->connections[$slice] = new Connection($config, new Driver());
```

Comment:
I wonder if it's worth running a `USE ${db}_p` query here? Otherwise the user has to add the DB name to their SQL. (Or, maybe not here, but at least before returning from this method.)

Reply:
Done by passing in the `dbname` key in the config hash. Of course this does not change when you get the connection for an additional database on the same slice (since they share the same connection). If the developer intends on communicating with more than one db, I suppose they'll have to always specify it in their queries since there are shared connections, but this is how it is normally done for multi-db applications anyway.